hazelcast/src/main/java/com/hazelcast/jet/impl/connector/HazelcastWriters.java

```diff
@@ -195,6 +195,11 @@ public int preferredLocalParallelism() {
             public Permission getRequiredPermission() {
                 return new RingBufferPermission(name, ACTION_CREATE, ACTION_PUT);
             }
+
+            @Override
+            public boolean isStateful() {
```

Comment:
there are only a few stateless PMSes, most notable being `HazelcastWriters`. If almost any non-trivial DAG (in particular for SQL) can use this optimization, then maybe it does not make sense?

Reply:
In particular, you assume that `MetaSupplierFromProcessorSupplier` (used by many SQL operators, but not only) can be stateful. `ProcessorMetaSupplier` javadoc states that ProcessorSupplier is always serialized, but possibly it is not for local member, am I correct? BTW, that javadoc should be updated because we not always serialize PMS

Reply:
We can add `isStateful()` method to `ProcessorSupplier`, so that `MetaSupplierFromProcessorSupplier` doesn’t have to return true always. We can further optimize this by adding `isStateful()` to `Processor` as well. But, in the end, `ConvenientSourceP`, for example, has to return true since it cannot know whether `fillBufferFn` is stateful in advance. What the problem here is that the technical definition of “statefulness” is not suitable for our purposes. Processors might have internal state, which makes them stateful by definition, but as long as they don't reach/update external variables, they don't need the be serialized. We can be more direct by adding something like `needsSerialization()`.

Reply:
`Processor` is one step too far. Processors are usually stateful, but that is not a problem. We are interested in what has state during `Processor`s creation. And the processor does not have to know how it is created (apart from constructor parameters of course) - this is an implementation of PS, same applies to PMS.

> We can add isStateful() method to ProcessorSupplier, so that MetaSupplierFromProcessorSupplier doesn’t have to return true always
This could solve some cases, but the biggest problem is with lambdas. 
Also I can imagine any combination of PMS.isStateful() and PS.isStateful(). For example, it depends which of them creates heavy, shared objects.

Reply:
I increased the coverage of the optimization by assuming that `FunctionEx`, `SupplierEx`, `ProcessorSupplier` and `ProcessorMetaSupplier` are stateless by default. Currently, only Elasticsearch, Hadoop and Kinesis PS/PMS are stateful due to mutable configuration. However, I am doubtful about the usefulness of the optimization after realizing that `ProcessorSupplier` is serialized for the local member (for [normal][1] / [light][2] jobs). Maybe we should optimize that as well?

[1]: https://github.com/burakgok/hazelcast/blob/41c2f0f228f25ebe697c4d8bcbcac7b78cbc4325/hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java#L297
[2]: https://github.com/burakgok/hazelcast/blob/41c2f0f228f25ebe697c4d8bcbcac7b78cbc4325/hazelcast/src/main/java/com/hazelcast/jet/impl/LightMasterContext.java#L160

Reply:
The default value for `ProcessorMetaSupplier` must be `false` if we want to keep/fix our the backwards compatibility guarantees.

If a user had a custom PMS which worked ok until 5.2.1, then it broke. With this PR such PMS should start working again, without the user having to override this method.

Reply:
> If a user had a custom PMS which worked ok until 5.2.1, then it broke.

We have only one bug report for this optimization, in which a CSV file is read from AWS S3. The functionality is internally implemented by Hadoop processors and this PR addresses that issue. If you are aware of other incidinces, please share them. The API says that PMS is always serialized if the job is submitted via a client, but doesn't clarify the serialization behavior for light jobs when the job is submitted via a member. We cannot be blamed for breaking a promise which we didn't make. `ProcessorMetaSupplier`s are not expected to be stateful and in Elastic/Hadoop/Kinesis PMSs, the root cause of the problem is their mutable configurations. As a solution, we can even simply clone the configurations; we don't need the mechanism I propose in this PR. By making it clear in the API that PMS should be stateless, we can ensure that implementors will not encounter the same problem in the future as long as they read the API.

Reply:
>The API says that PMS is always serialized if the job is submitted via a client, but doesn't clarify the serialization behavior for light jobs when the job is submitted via a member. We cannot be blamed for breaking a promise which we didn't make

We did not foresee the optimisation opportunity. IMO PMS javadoc is clear: PMS is always serialized and PS is always serialized, because there are no exceptions. PMS and PS behavior is the same for normal and light jobs. I think that `client` is mentioned only because the doc is very old and predates both light jobs and SQL that creates jobs on members.

We should update PMS javadoc as part of this PR.

Reply:
BTW, `init` javadoc in PMS: 
> Called on the cluster member that receives the client request, **after deserializing** the meta-supplier instance

and in PS:
> Called on each cluster member **after deserialization**

Reply:
> By making it clear in the API that PMS should be stateless

That would be a breaking change, but IMO might have sense. Are there any cases where PMS or PS must be stateful?